File: apps/mp/api/vip.ts
/**
 * VIP 域只读 API 客户端 (FE-S3-004 + FE-S3-005).
 *
 * 对齐后端 ``apps/api/app/api/v1/vip.py`` (BE-S3-009):
 * - GET ``/api/v1/vip/me``      当前用户订阅状态 (auth required)
 * - GET ``/api/v1/vip/orders``  订单历史最近 N 条 (auth required, 默认 20, 最大 100)
 *
 * 字段 100% 对齐 ``apps/api/app/schemas/vip.py``, 不做 alias / 缩写.
 *
 * 设计取舍:
 *
 * - **vip 与 payment 拆两个文件**: vip 是会员状态读路径 (展示用), payment 是
 *   商业化交易写路径 (下单 / 回调). 与后端模块拆分一致, 减少互相 import
 *
 * - **``MembershipResponse.has_active=false`` 时其他字段仍可能有值**: 历史信息 ──
 *   "用户曾经有订阅但已 expired", UI 用来决定 "重新订阅 / 立即续费" 文案;
 *   完全 NULL 表示从未有过订阅记录 (``vip_trial_days=0`` 注册的用户 + 从未付费)
 *
 * - **不暴露 ``raw_callback``**: BE 已在 ``OrderResponse`` 显式 select 字段, 微信回调
 *   原始 payload 含 PII + 商户敏感数据, 不进前端
 */

import { request } from '@/utils/request'

export type VipStatus = 'trialing' | 'active' | 'expired' | 'cancelled'
export type VipPlan = 'trial' | 'monthly' | 'quarterly' | 'yearly' | 'lifetime'
export type OrderStatus = 'pending' | 'paid' | 'failed' | 'refunded'
export type OrderPaymentChannel = 'wechat_mp' | 'wechat_h5' | 'apple_iap' | 'internal'

export interface MembershipResponse {
  has_active: boolean
  membership_id: string | null
  user_id: string
  status: VipStatus | null
  plan: VipPlan | null
  /** ISO8601 字符串 */
  start_at: string | null
  end_at: string | null
  auto_renew: boolean
  /** 累计支付 CNY (Decimal 序列化为 number) */
  total_paid_cny: number
  /** 距离 end_at 剩余天数 (向下取整). lifetime 返 36500+ 大数; 无订阅 / expired 返 null */
  days_remaining: number | null
}

export interface OrderItem {
  order_id: string
  out_trade_no: string
  plan: VipPlan
  amount_cny: number
  status: OrderStatus
  payment_channel: OrderPaymentChannel
  transaction_id: string | null
  /** 微信回调 success_time (ISO8601); pending / failed 为 null */
  paid_at: string | null
  created_at: string
}

export interface OrdersListResponse {
  items: OrderItem[]
  total: number
}

/**
 * 拉当前用户订阅状态 (BE-S3-009 ``GET /vip/me``).
 *
 * 单点查 user_id UNIQUE 索引, 端到端 < 5ms. 调用场景:
 * - 个人中心 onShow (展示 VIP 卡)
 * - 支付结果页 onLoad (轮询 / 一次性确认 active)
 * - auth.refreshMembership() 集中入口
 */
export function fetchMembership() {
  return request<MembershipResponse>({
    url: '/api/v1/vip/me',
    method: 'GET',
  })
}

/**
 * 拉当前用户订单历史 (BE-S3-009 ``GET /vip/orders``).
 *
 * 倒序 (created_at DESC) 走 ix_vip_orders_user_created 索引. 不分页, 默认 20 条
 * (lifetime + 长期续费用户 ≤ 100 笔, FE-S3-005 订单历史页用).
 */
export function fetchOrders(limit = 20) {
  return request<OrdersListResponse>({
    url: `/api/v1/vip/orders?limit=${limit}`,
    method: 'GET',
  })
}

File: apps/api/app/schemas/vip.py
"""VIP 域 Pydantic 模型 (BE-S3-009).

3 个面向客户端的响应:
- ``MembershipResponse``  ← ``GET /vip/me``     (查自己的订阅状态)
- ``OrderResponse``       ← ``GET /vip/orders`` (订单列表项)
- ``OrdersListResponse``  ← ``GET /vip/orders`` (订单列表 + 分页元信息)

ORM 字段 → Schema 字段 100% 对齐 (大小写 + 命名), 不做 alias / 缩写,
避免前端 / 后端各自猜命名 (减少跨端 contract bug).
"""

from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

PlanLiteral = Literal["trial", "monthly", "quarterly", "yearly", "lifetime"]
StatusLiteral = Literal["trialing", "active", "expired", "cancelled"]
PaymentChannelLiteral = Literal[
    "wechat_mp", "wechat_h5", "apple_iap", "internal"
]
OrderStatusLiteral = Literal["pending", "paid", "failed", "refunded"]


class MembershipResponse(BaseModel):
    """``GET /vip/me`` 返回; 当前用户订阅信息.

    用户从未付费 / 注册但 ``vip_trial_days=0`` 时返回所有字段为 None 的"伪 membership".
    """

    model_config = ConfigDict(extra="forbid", from_attributes=True)

    has_active: bool = Field(
        description="是否当前生效 (status active/trialing + end_at > now). False 时其它字段仍可能有值 (历史订阅信息)."
    )
    membership_id: uuid.UUID | None = Field(
        default=None, description="None 表示用户从未有过订阅记录"
    )
    user_id: uuid.UUID
    status: StatusLiteral | None = Field(default=None)
    plan: PlanLiteral | None = Field(default=None)
    start_at: datetime | None = Field(default=None)
    end_at: datetime | None = Field(default=None)
    auto_renew: bool = Field(default=False)
    total_paid_cny: Decimal = Field(
        default=Decimal("0"),
        description="累计支付 CNY; 试用期间 = 0; 续费时累加 (财务对账依据)",
    )
    days_remaining: int | None = Field(
        default=None,
        ge=0,
        description="距离 end_at 剩余天数 (向下取整). lifetime 返 36500+ 大数; expired 返 None.",
    )


class OrderResponse(BaseModel):
    """``GET /vip/orders`` 列表项.

    ``raw_callback`` (微信回调原始 payload) 不暴露给客户端 (PII + 商户敏感数据).
    """

    model_config = ConfigDict(extra="forbid", from_attributes=True)

    order_id: uuid.UUID
    out_trade_no: str
    plan: PlanLiteral
    amount_cny: Decimal
    status: OrderStatusLiteral
    payment_channel: PaymentChannelLiteral
    transaction_id: str | None = None
    paid_at: datetime | None = None
    created_at: datetime


class OrdersListResponse(BaseModel):
    """``GET /vip/orders`` 顶层返回 (含 limit + total 元数据).

    暂不分页 (用户订单 < 100 条; lifetime 上限 ≤ 20 笔), 直接返全量.
    """

    model_config = ConfigDict(extra="forbid")

    items: list[OrderResponse]
    total: int = Field(ge=0, description="返回项数 = items 长度 (一次取全)")


__all__ = [
    "MembershipResponse",
    "OrderResponse",
    "OrdersListResponse",
    "OrderStatusLiteral",
    "PaymentChannelLiteral",
    "PlanLiteral",
    "StatusLiteral",
]
